pkg/controllers/securitypolicy_associated_controller.go

```diff
@@ -0,0 +1,250 @@
+			shouldReconcile = true
+			break
+		}
+	}
+	if !shouldReconcile {
+		log.Info("no pod in namespace is relevant", "namespace", obj.Namespace, "name", obj.Name)
+		return
+	}
+
+	err = reconcileSecurityPolicy(e.Client, podList.Items, l)
+	if err != nil {
+		log.Error(err, "failed to reconcile security policy")
+	}
+}
+
+type EnqueueRequestForPod struct {
+	Client client.Client
+}
+
+func (e *EnqueueRequestForPod) Create(createEvent event.CreateEvent, q workqueue.RateLimitingInterface) {
+	e.Raw(createEvent, q)
+}
+
+func (e *EnqueueRequestForPod) Update(updateEvent event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	e.Raw(updateEvent, q)
+}
+
+func (e *EnqueueRequestForPod) Delete(deleteEvent event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	e.Raw(deleteEvent, q)
+}
+
+func (e *EnqueueRequestForPod) Generic(genericEvent event.GenericEvent, q workqueue.RateLimitingInterface) {
+	e.Raw(genericEvent, q)
+}
+
+func (e *EnqueueRequestForPod) Raw(evt interface{}, q workqueue.RateLimitingInterface) {
+	var pods []v1.Pod
+	var obj client.Object
+
+	switch et := evt.(type) {
+	case event.CreateEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.UpdateEvent:
+		// pod doesn't support port update alone, it will delete and create a new pod instead.
+		return
+	case event.DeleteEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.GenericEvent:
+		obj = et.Object.(*v1.Pod)
+	default:
+		log.Error(nil, "unknown event type", "event", evt)
+	}
+	pod := obj.(*v1.Pod)
+	if isInSysNs, err := isNsInSystemNamespace(e.Client, pod.Namespace); err != nil {
+		return
+	} else if isInSysNs {
+		log.Info("pod is in system namespace, do nothing")
+		return
+	}
+	err := reconcileSecurityPolicy(e.Client, pods, q)
+	if err != nil {
+		log.Error(err, "failed to reconcile security policy")
+	}
+}
+
+func getAllPodPortNames(pods []v1.Pod) sets.String {
+	podPortNames := sets.NewString()
+	for _, pod := range pods {
+		for _, container := range pod.Spec.Containers {
+			for _, port := range container.Ports {
+				if port.Name != "" {
+					podPortNames.Insert(port.Name)
+				}
+			}
+		}
+	}
+	return podPortNames
+}
+
+func reconcileSecurityPolicy(client client.Client, pods []v1.Pod, q workqueue.RateLimitingInterface) error {
+	podPortNames := getAllPodPortNames(pods)
+	log.V(1).Info("pod named port", "podPortNames", podPortNames)
+	spList := &v1alpha1.SecurityPolicyList{}
+	err := client.List(context.Background(), spList)
+	if err != nil {
+		log.Error(err, "failed to list security policy")
+		return err
+	}
+
+	for _, securityPolicy := range spList.Items {
+		shouldReconcile := false
+		for _, rule := range securityPolicy.Spec.Rules {
+			for _, port := range rule.Ports {
+				if port.Port.Type == intstr.String {
+					if podPortNames.Has(port.Port.StrVal) {
+						shouldReconcile = true
+					}
+				}
+			}
+		}
+		if shouldReconcile {
+			log.Info("reconcile security policy", "namespace", securityPolicy.Namespace, "name", securityPolicy.Name)
```

Comment:
reconcile security policy because of named port change

Reply:
ok